WcaOnRails/app/models/competition.rb

```diff
@@ -140,10 +146,44 @@ def info_for(user)
     info
   end
 
+  attr_accessor :cloned_from_id
+  def cloned_from
+    Competition.find_by(id: cloned_from_id)
+  end
+
   def build_clone
     Competition.new(attributes.slice(*CLONEABLE_ATTRIBUTES)).tap do |clone|
-      clone.organizers = organizers
-      clone.delegates = delegates
+      clone.cloned_from_id = id
+
+      Competition.reflections.keys.each do |association_name|
+        case association_name
+        when 'registrations', 'results', 'competitors', 'competitor_users', 'delegate_report',
+             'competition_delegates', 'competition_organizers', 'media', 'scrambles'
+          # Should be neither cloned nor copied.
+        when 'organizers'
+          clone.organizers = organizers
+        when 'delegates'
+          clone.delegates = delegates
+        when 'competition_tabs'
+          # Clone tabs in the clone_associations callback after the competition is saved.
+          clone.clone_tabs = true # True by default.
+        else
+          raise "Cloning/copying behavior for Competition.#{association_name} is not defined. See Competition#build_clone."
+        end
+      end
+    end
+  end
+
+  attr_accessor :clone_tabs
+
+  # After the cloned competition is created, clone other associations which cannot just be copied.
+  after_create :clone_associations
+  private def clone_associations
+    # Clone competition_tabs.
+    if clone_tabs
+      cloned_from&.competition_tabs.each do |tab|
```

Comment:
This should either be `cloned_from.competition_tabs.each` or `cloned_from&.competition_tabs&.each`. If `cloned_from` is nil, `cloned_from&.competition_tabs` will be nil, and `.each` will blow up.

I prefer the confident latter design, but this is hardly the place for that philosophical discussion. Pick either one :)